cocotb/types/logic.py

```diff
@@ -0,0 +1,279 @@
+        None: 2,
+        "X": 2,
+        "x": 2,
+        "U": 2,
+        "u": 2,
+        "W": 2,
+        "w": 2,
+        "-": 2,
+        # high impedance
+        "Z": 3,
+        "z": 3,
+    }
+
+    @lru_cache(maxsize=None)
+    def __new__(cls, value: Optional[Any] = None) -> "Logic":
+        # convert to internal representation
+        try:
+            _repr = cls._repr_map[value]
+        except KeyError:
+            raise ValueError(
+                "{!r} is not convertible to a {}".format(value, cls.__qualname__)
+            ) from None
+        obj = super().__new__(cls)
+        obj._repr = _repr
+        return obj
+
+    def __and__(self, other: "Logic") -> "Logic":
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return type(self)(
+            (
+                ("0", "0", "0", "0"),
+                ("0", "1", "X", "X"),
+                ("0", "X", "X", "X"),
+                ("0", "X", "X", "X"),
+            )[self._repr][other._repr]
+        )
+
+    def __rand__(self, other: "Logic") -> "Logic":
+        return self & other
+
+    def __or__(self, other: "Logic") -> "Logic":
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return type(self)(
+            (
+                ("0", "1", "X", "X"),
+                ("1", "1", "1", "1"),
+                ("X", "1", "X", "X"),
+                ("X", "1", "X", "X"),
+            )[self._repr][other._repr]
+        )
+
+    def __ror__(self, other: "Logic") -> "Logic":
+        return self | other
+
+    def __xor__(self, other: "Logic") -> "Logic":
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return type(self)(
+            (
+                ("0", "1", "X", "X"),
+                ("1", "0", "X", "X"),
+                ("X", "X", "X", "X"),
+                ("X", "X", "X", "X"),
+            )[self._repr][other._repr]
+        )
+
+    def __rxor__(self, other: "Logic") -> "Logic":
+        return self ^ other
+
+    def __invert__(self) -> "Logic":
+        return type(self)(("1", "0", "X", "X")[self._repr])
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return self._repr == other._repr
+
+    def __hash__(self) -> int:
+        return self._repr
+
+    def __repr__(self) -> str:
+        return "{}({!r})".format(type(self).__qualname__, str(self))
+
+    def __str__(self) -> str:
+        return ("0", "1", "X", "Z")[self._repr]
+
+    def __bool__(self) -> bool:
+        if self._repr < 2:
+            return bool(self._repr)
+        raise ValueError(f"Cannot convert {self!r} to bool")
+
+    def __int__(self) -> int:
+        if self._repr < 2:
+            return self._repr
+        raise ValueError(f"Cannot convert {self!r} to int")
+
+
+class Bit(Logic):
+    r"""
+    Model of a 2-value (``0``, ``1``) datatype commonly seen in HDLs.
```

Comment:
A remark somewhere that explains that `Bit` is a stricter version of the builtin `bool` that does not allow arithmetic and prints as 0/1 instead of `True/False` would be a good thing.